plugins/kubernetes-backend/src/cluster-locator/MultiTenantConfigClusterLocator.ts

```diff
@@ -28,20 +29,45 @@ export class MultiTenantConfigClusterLocator
   }
 
   static fromConfig(config: Config[]): MultiTenantConfigClusterLocator {
+    // TODO: Add validation that authProvider is required and serviceAccountToken
+    // is required if authProvider is serviceAccount
     return new MultiTenantConfigClusterLocator(
       config.map(c => {
         return {
           name: c.getString('name'),
           url: c.getString('url'),
           serviceAccountToken: c.getOptionalString('serviceAccountToken'),
+          authProvider: (c.getOptionalString('authProvider')
+            ? c.getOptionalString('authProvider')
+            : 'serviceAccount') as AuthProviderType,
         };
       }),
     );
   }
 
   // As this implementation always returns all clusters serviceId is ignored here
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  async getClusterByServiceId(_serviceId: string): Promise<ClusterDetails[]> {
-    return this.clusterDetails;
+  async getClusterByServiceId(
+    _serviceId: string,
+    authTokens: AuthTokens,
+  ): Promise<ClusterDetails[]> {
+    return this.clusterDetails.map(cd => {
+      const authProviderKind = cd.authProvider;
+      if (authProviderKind === 'google') {
+        const authToken: string | undefined = authTokens[authProviderKind];
+        const clusterDetailWithAuthToken: ClusterDetails = Object.assign(
+          {},
+          cd,
+        );
+        clusterDetailWithAuthToken.serviceAccountToken = authToken;
+        return clusterDetailWithAuthToken;
+        // TODO: When validation for authProvider key is added, remove allowance of undefined as value
+      } else if (authProviderKind === 'serviceAccount') {
+        return cd;
+      }
+      throw new Error(
+        `Unsupported Kubernetes auth provider "${authProviderKind}"`,
+      );
+    });
```

Comment:
I'm a bit concerned that we're doing too much here. Possibly this method should return the clusterDetails without tokens (except those from the config) and we combine them in a separate method.

This will probably be necessary anyway when other `ClusterLocators` are brought in (for example the HTTP get one).

Reply:
Yeah definitely need to invert the responsibilities a bit here.

We also need to make sure it's possible to provide custom auth providers for organizations with their own setup or unique way to auth towards different clusters. Ideally it's a TypeScript interface that we provide a couple of implementations of in this project, but allow custom implementations to be passed in too.